numba/npyufunc/array_exprs.py

```diff
@@ -0,0 +1,349 @@
+                ir_op))
+
+    def _translate_expr(self, ir_expr):
+        '''Translate the given expression from Numba IR to an array expression
+        tree.
+        '''
+        ir_op = ir_expr.op
+        if ir_op == 'arrayexpr':
+            return ir_expr.expr
+        operands_or_args = [self.const_assigns.get(op_var.name, op_var)
+                            for op_var in self._get_operands(ir_expr)]
+        return self._get_array_operator(ir_expr), operands_or_args
+
+    def _handle_matches(self):
+        '''Iterate over the matches, trying to find which instructions should
+        be rewritten, deleted, or moved.
+        '''
+        replace_map = {}
+        dead_vars = set()
+        used_vars = set()
+        for match in self.matches:
+            instr = self.array_assigns[match]
+            expr = instr.value
+            arr_inps = []
+            arr_expr = self._get_array_operator(expr), arr_inps
+            new_expr = ir.Expr(op='arrayexpr',
+                               loc=expr.loc,
+                               expr=arr_expr,
+                               ty=self.typemap[instr.target.name])
+            new_instr = ir.Assign(new_expr, instr.target, instr.loc)
+            replace_map[instr] = new_instr
+            self.array_assigns[instr.target.name] = new_instr
+            for operand in self._get_operands(expr):
+                operand_name = operand.name
+                if operand_name in self.array_assigns:
+                    child_assign = self.array_assigns[operand_name]
+                    child_expr = child_assign.value
+                    child_operands = child_expr.list_vars()
+                    used_vars.update(operand.name
+                                     for operand in child_operands)
+                    arr_inps.append(self._translate_expr(child_expr))
+                    if child_assign.target.is_temp:
+                        dead_vars.add(child_assign.target.name)
+                        replace_map[child_assign] = None
+                elif operand_name in self.const_assigns:
+                    arr_inps.append(self.const_assigns[operand_name])
+                else:
+                    used_vars.add(operand.name)
+                    arr_inps.append(operand)
+        return replace_map, dead_vars, used_vars
+
+    def _get_final_replacement(self, replacement_map, instr):
+        '''Find the final replacement instruction for a given initial
+        instruction by chasing instructions in a map from instructions
+        to replacement instructions.
+        '''
+        replacement = replacement_map[instr]
+        while replacement in replacement_map:
+            replacement = replacement_map[replacement]
+        return replacement
+
+    def apply(self):
+        '''When we've found array expressions in a basic block, rewrite that
+        block, returning a new, transformed block.
+        '''
+        if config.DUMP_IR:
```

Comment:
Instead of duplicating this in each rewrite pass, we probably want to put it in the RewriteRegistry, no?